# Analisis Exploratorio de Datos del dataset PlantVillage
Nombres: 
- Alejandro Reyes V.
- Felipe Sánchez C.
___

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf


I0000 00:00:1784321227.271906   25283 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
I0000 00:00:1784321227.335162   25283 cpu_feature_guard.cc:227] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
I0000 00:00:1784321228.836274   25283 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.


## Estructura del Dataset:

Se trabajará con el dataset PlantVillage de 'spMoganty', el cual se puede obtener desde el [repositorio oficial en github](https://github.com/spMohanty/PlantVillage-Dataset.git) o desde [su pagina en Kaggle](https://www.kaggle.com/datasets/mohitsingh1804/plantvillage)

Este dataset contiene 38 tipos de plantas, con un total de 54305 imágenes que incluyen plantas sanas y con pestes. En total, el dataset cubre 14 especies de cultivos, entre los cuales se encuentran manzanos, naranjos, cerezos, entre otros.

Para efectos de este proyecto, los cultivos relevantes son Cerezos, que en el dataset se separan en 2 carpetas (sanos y con peste), y que a su vez se distribuyen en conjuntos de entrenamiento y validación (train/ y val/).

In [2]:

!ls cherry_disease/train/
!ls cherry_disease/val/



'Cherry_(including_sour)___healthy'  'Cherry_(including_sour)___Powdery_mildew'
'Cherry_(including_sour)___healthy'  'Cherry_(including_sour)___Powdery_mildew'


In [8]:
import os

# Dimensiones del dataset

blanquilla_num_train = len(os.listdir('./cherry_disease/train/Cherry_(including_sour)___Powdery_mildew'))
sana_num_train = len(os.listdir('./cherry_disease/train/Cherry_(including_sour)___healthy'))
total = blanquilla_num_train + sana_num_train

blanquilla_num_val = len(os.listdir('./cherry_disease/val/Cherry_(including_sour)___Powdery_mildew'))
sana_num_val = len(os.listdir('./cherry_disease/val/Cherry_(including_sour)___healthy'))
total_val = blanquilla_num_val + sana_num_val

print("Número total de imágenes en el dataset(train):", total)
print("Número de imágenes de hojas con blanquilla:", blanquilla_num_train, "({:.2f}%)".format((blanquilla_num_train / total) * 100))
print("Número de imágenes de hojas sanas:", sana_num_train, "({:.2f}%)".format((sana_num_train / total) * 100))
print("----------------------------------------------------------------")
print("Número total de imágenes en el dataset(validation):", total_val)
print("Número de imágenes de hojas con blanquilla:", blanquilla_num_val, "({:.2f}%)".format((blanquilla_num_val / total_val) * 100))
print("Número de imágenes de hojas sanas:", sana_num_val, "({:.2f}%)".format((sana_num_val / total_val) * 100))

Número total de imágenes en el dataset(train): 1526
Número de imágenes de hojas con blanquilla: 842 (55.18%)
Número de imágenes de hojas sanas: 684 (44.82%)
----------------------------------------------------------------
Número total de imágenes en el dataset(validation): 380
Número de imágenes de hojas con blanquilla: 210 (55.26%)
Número de imágenes de hojas sanas: 170 (44.74%)


### Distribución del dataset

Desde lo anterior obtenemos que la totalidad del conjunto de entrenamiento es de 1526 imágenes. De las cuales el 55.2% corresponde a hojas con blanquilla, y el 44.8% a hojas sanas.
Esto demuestra que las clases (sana / enferma) estan balanceadas.

Por otro lado, el conjunto de validacion incluye un total de 380 imágenes, 55.3% de hojas con peste y 44.7% de hojas sanas.

Lo anterior indica que el dataset sigue una proporcion de 80/20 en training y validation.


Para efectos de este proyecto, y para evitar resultados pre-fabricados, se realizará un shuffle entre el conjunto de entrenamiento y de validación, manteniendo las proporciones.

In [ ]:
# shuffle entre train y val, manteniendo la proporción de clases

train_directory = 'cherry_disease/train'
val_directory = 'cherry_disease/val'









# train_dataset = tf.keras.utils.image_dataset_from_directory(
#     train_directory,
#     labels='inferred',       # infiere la etiqueta desde el nombre de la carpeta
#     label_mode='binary',     # usar binary pues se tiene hojas sanas y enfermas
#     image_size=(224, 224),   # resize
#     batch_size=32            # # de imágenes procesadas a la vez
# )

Found 1526 files belonging to 2 classes.


In [10]:
#Clases detectadas en el dataset

nombres_clases = train_dataset.class_names
print("Clases detectadas:", nombres_clases)

Clases detectadas: ['Cherry_(including_sour)___Powdery_mildew', 'Cherry_(including_sour)___healthy']
